src/Microsoft.AspNetCore.Mvc.Modules/ModularApplicationPart.cs

```diff
@@ -74,18 +75,23 @@ public IEnumerable<TypeInfo> Types
         {
             get
             {
-                return DefaultAssemblyDiscoveryProvider
-                    .DiscoverAssemblies(
-                        ShellBlueprint.Dependencies.Select(dep => dep.Type.GetTypeInfo().Assembly).Distinct().ToList(),
-                        ReferenceAssemblies)
+                var assemblies = ModularAssemblyProvider
+                    .GetAssemblies(ExtensionLibraryService.RuntimeLibraries);
+
+                var types = DefaultModularAssemblyDiscoveryProvider
+                    .GetCandidateAssemblies(assemblies, ReferenceAssemblies)
                     .SelectMany(x => x.DefinedTypes);
+
+                return types;
             }
         }
 
         /// <inheritdoc />
         public IEnumerable<string> GetReferencePaths()
         {
-            return AssemblyProvider.GetAssemblies().Select(x => x.Location);
+            return DependencyContext.Default.CompileLibraries
```

Comment:
>I think the modules should have preservecompilationcontext on them....

Good idea, so we would need to use this option for each module and use a dependency context on it. Just re-tried, the module assembly is bigger and more assemblies are added in its bin folder. It works but paths are not fully resolved. Currently this is done through the project model in `ExtensionLibraryService` which provides `ReferencePaths`.

>Use the ShellDescriptor to get the assemblies from the types.

Yes, you're right. Notice that for Types i now use as you the `shellBlueprint` through `ModularAssemblyProvider`. Here, it's just to show that we not only need the compilation references of modules but also of all their dependencies. Those related to the ambient libraries are provided by `DependencyContext.Default`, other ones by our `ReferencePaths`.

Note: `ExtensionLibraryService` could do this by not limiting references to non ambient libraries, and returning a dictionary of references lists for each module. Then we could filter by modules / features.

But right now, the most references we have are for the ambient framework compilation assemblies that we need for razor views compilation purpose. We don't need them in the master branch because `AddMvcCore()` is done at the app level. Tomorrow, maybe we will not need it here too if we reference all modules through the main app.

>Use DependencyContext.Load(assemblies) - use context.CompileLibraries

Yes, would be good but we can't use it without `preservecompilationcontext` for each modules, and, even with this option, references are not fully resolved (see above). This only works for the main app compilation assemblies.

>Make sure all is distinct

For non ambient libraries, our `ReferencePaths` are already distinct, and then they are distinct from the ones grabbed from the main `DependencyContext.Default`. Note that to use this last one to grab compilation references you need at least `preservecompilationcontext` for `Orchard.Cms.Web`. Otherwise you can only resolve runtimes ones.